File: solutions/javascript/mixed-juices/4/mixed-juices.js
// @ts-check

const JUICES_TO_MINUTES = {
    'All or Nothing' : 5,
    'Energizer': 1.5,
    'Green Garden': 1.5,
    'Pure Strawberry Joy': 0.5,
    'Tropical Island': 3,
};

const LIME_SIZES_TO_WEDGES = {
    small: 6,
    medium: 8,
    large: 10,
};

/**
 * Determines how long it takes to prepare a certain juice.
 *
 * @param {string} name
 * @returns {number} time in minutes
 */
export function timeToMixJuice(name) {
    return JUICES_TO_MINUTES[name] || 2.5;
}

/**
 * Calculates the number of limes that need to be cut
 * to reach a certain supply.
 *
 * @param {number} wedgesNeeded
 * @param {string[]} limes
 * @returns {number} number of limes cut
 */
export function limesToCut(wedgesNeeded, limes) {
    let limesCut = 0;

    while (wedgesNeeded > 0 && limes.length) {
        wedgesNeeded -= LIME_SIZES_TO_WEDGES[limes.shift()];
        limesCut++;
    }

    return limesCut;
}

/**
 * Determines which juices still need to be prepared after the end of the shift.
 *
 * @param {number} timeLeft
 * @param {string[]} orders
 * @returns {string[]} remaining orders after the time is up
 */
export function remainingOrders(timeLeft, orders) {
    while (timeLeft > 0 && orders.length) {
        timeLeft -= timeToMixJuice(orders.shift());
    }

    return orders;
}

File: solutions/julia/mixed-juices/6/mixed-juices.jl
function time_to_mix_juice(juice)
    return get(Dict(
                    "Pure Strawberry Joy" => 0.5,
                    "Energizer" => 1.5,
                    "Green Garden" => 1.5,
                    "Tropical Island" => 3.0,
                    "All or Nothing" => 5.0,
                   ), juice, 2.5)
end

function wedges_from_lime(size)
    size == "small" && return 6
    size == "medium" && return 8
    size == "large" && return 10
end

function limes_to_cut(needed, limes)
    limes_cut = 0

    while needed > 0
        length(limes) == 0 && break
        needed -= wedges_from_lime(popfirst!(limes))
        limes_cut += 1
    end

    return limes_cut
end

order_times(orders) = [time_to_mix_juice(order) for order in orders]

function remaining_orders(time_left, orders)
    while time_left > 0
        length(orders) == 0 && break
        time_left -= time_to_mix_juice(popfirst!(orders))
    end

    return orders
end
